### ISSUE TITLE
false negative with dangerous-default-value

### ISSUE THREAD

**Issue by dpinol** [pylint-dev/pylint#8826]
### Bug description

```python
# pylint: disable=missing-module-docstring
from dataclasses import dataclass
# pylint: disable=missing-class-docstring
# pylint: disable=too-few-public-methods
class Arg:
    def __init__(self):
        self.field = 4


@dataclass(frozen=True)
class Frozen:
    field: int = 3


# pylint: disable=missing-function-docstring
def func(arg=Arg()):
    arg.field = 5


# pylint: disable=missing-function-docstring
def func2(frozen=Frozen()):
    print(frozen.field)
```

### Command used

```shell
pylint kk.py
```

### Pylint output

```shell
-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.00/10, +2.00)
```

### Expected behavior

ttbomk it should report `dangerous-default-value` for func, but not for func2 since the argument is readonly.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.5
Python 3.11.3 (main, May  2 2023, 16:42:49) [GCC 11.3.0]
```

### OS / Environment

ubuntu 22.10

**Comment by jacobtylerwalls** [comment]
Thanks for the report! IMO this is a duplicate of #4659. Please feel free to review the PR for it and offer a view on moving it forward.

**Comment by dpinol** [comment]
Hi, thanks for the quick answer.
IMO #4659 is different because it refers to immutable stateful objects. 
What I mean is that [dangerous-default-value](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/dangerous-default-value.html) refers to "a mutable value as list or dictionary". If "dangerous-default-value" is only meant for dicts and lists, I think the documentation should be more explicit to say instead  "list or dictionary"

**Comment by jacobtylerwalls** [comment]
> IMO https://github.com/pylint-dev/pylint/issues/4659 is different because it refers to immutable stateful objects.

As I read #4659 it was suggesting collapsing the distinction between "mutable" and "stateful", not introducing a distinctive "immutable stateful" class.

In other words, suggesting expanding the message beyond dicts and lists to callables that we can reasonably detect return something other than a constant. Here, you have a callable `Arg` returning a dataclass instance, which isn't a constant. A frozen dataclass would be like a constant, so I left a comment to that effect.

> If "dangerous-default-value" is only meant for dicts and lists, I think the documentation should be more explicit to say instead "list or dictionary"

Did you mean if "is not only meant"? Yes, the docs would need to be updated as part of that PR.
